include/libopencm3/swm050/gpio.h

```diff
@@ -0,0 +1,86 @@
+/*
+ * This file is part of the libopencm3 project.
+ *
+ * Copyright (C) 2019 Icenowy Zheng <icenowy@aosc.io>
+ *
+ * This library is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef LIBOPENCM3_GPIO_H
+#define LIBOPENCM3_GPIO_H
+
+#include <libopencm3/cm3/common.h>
+#include <libopencm3/swm050/memorymap.h>
+
+/* GPIO number definitions (for convenience) */
+/** @defgroup gpio_pin_id GPIO Pin Identifiers
+@ingroup gpio_defines
+
+@{*/
+#define GPIO0				(1 << 0)
+#define GPIO1				(1 << 1)
+#define GPIO2				(1 << 2)
+#define GPIO3				(1 << 3)
+#define GPIO4				(1 << 4)
+#define GPIO5				(1 << 5)
+#define GPIO6				(1 << 6)
+#define GPIO7				(1 << 7)
+#define GPIO8				(1 << 8)
+#define GPIO9				(1 << 9)
+#define GPIO_ALL			0x3ff
+/**@}*/
+
+/* GPIO direction definitions */
+/** @defgroup gpio_dir GPIO Pin Direction
+@ingroup gpio_defines
+@{*/
+#define GPIO_INPUT			0x0
+#define GPIO_OUTPUT			0x1
+/**@}*/
+
+#define GPIO_DATA			MMIO32(GPIO_BASE + 0x0)
+#define GPIO_DIR			MMIO32(GPIO_BASE + 0x4)
+#define GPIO_EXT			MMIO32(GPIO_BASE + 0x4c)
```

Comment:
one last thing, which pdf did you get these values from?  I'm looking at and it has "PORTA_EXT" at 0x50, with "INTEOI" at 0x4c, and 0x48 as "reversed" (presumably reserved)

Reply:
[http://caxapa.ru/thumbs/870237/Foshan-Synwit-Tech-SWM050I2P7-65_C111818.pdf](url)

Reply:
I'm seeing the same thing

Reply:
Well, the summary table has the other address, the descriptions has the address as you have them... joy  I've kept it as  you've got here, but would love to see it confirmed, whether it matches the table or the descriptions in reality :)